pandas/tests/frame/test_reshape.py

```diff
@@ -136,6 +136,104 @@ def test_stack_unstack(self):
         assert_frame_equal(unstacked_cols.T, self.frame)
         assert_frame_equal(unstacked_cols_df['bar'].T, self.frame)
 
+    def test_unstack_fill(self):
+
+        # GH #9746: fill_value keyword argument for Series
+        # and DataFrame unstack
+
+        # From a series
+        data = Series([1, 2, 4, 5], dtype=np.int16)
+        data.index = MultiIndex.from_tuples(
+            [('x', 'a'), ('x', 'b'), ('y', 'b'), ('z', 'a')])
+
+        result = data.unstack(fill_value=-1)
+        expected = DataFrame({'a': [1, -1, 5], 'b': [2, 4, -1]},
+                             index=['x', 'y', 'z'], dtype=np.int16)
+        assert_frame_equal(result, expected)
+
+        # From a series with incorrect data type for fill_value
+        result = data.unstack(fill_value=0.5)
+        expected = DataFrame({'a': [1, 0.5, 5], 'b': [2, 4, 0.5]},
+                             index=['x', 'y', 'z'], dtype=np.float)
+        assert_frame_equal(result, expected)
+
+        # From a dataframe
+        rows = [[1, 2], [3, 4], [5, 6], [7, 8]]
+        df = DataFrame(rows, columns=list('AB'), dtype=np.int32)
+        df.index = MultiIndex.from_tuples(
+            [('x', 'a'), ('x', 'b'), ('y', 'b'), ('z', 'a')])
+
+        result = df.unstack(fill_value=-1)
+
+        rows = [[1, 3, 2, 4], [-1, 5, -1, 6], [7, -1, 8, -1]]
+        expected = DataFrame(rows, index=list('xyz'), dtype=np.int32)
+        expected.columns = MultiIndex.from_tuples(
+            [('A', 'a'), ('A', 'b'), ('B', 'a'), ('B', 'b')])
+        assert_frame_equal(result, expected)
+
+        # From a mixed type dataframe
+        df['A'] = df['A'].astype(np.int16)
+        df['B'] = df['B'].astype(np.float64)
+
+        result = df.unstack(fill_value=-1)
+        expected['A'] = expected['A'].astype(np.int16)
+        expected['B'] = expected['B'].astype(np.float64)
+        assert_frame_equal(result, expected)
+
+        # From a dataframe with incorrect data type for fill_value
+        result = df.unstack(fill_value=0.5)
+
+        rows = [[1, 3, 2, 4], [0.5, 5, 0.5, 6], [7, 0.5, 8, 0.5]]
+        expected = DataFrame(rows, index=list('xyz'), dtype=np.float)
+        expected.columns = MultiIndex.from_tuples(
+            [('A', 'a'), ('A', 'b'), ('B', 'a'), ('B', 'b')])
+        assert_frame_equal(result, expected)
+
+        # Test unstacking with date times
+        dv = pd.date_range('2012-01-01', periods=4).values
```

Comment:
Yes, timedeltas required some changes to _maybe_promote, good catch.  As
with datatime, cant seem to upcast to object.

On Wed, Jan 20, 2016 at 8:48 AM Jeff Reback <notifications@github.com>
wrote:

> In pandas/tests/frame/test_reshape.py
> <https://github.com/pydata/pandas/pull/10246#discussion_r50281164>:
>
> > +        result = df.unstack(fill_value=-1)
> > +        expected['A'] = expected['A'].astype(np.int16)
> > +        expected['B'] = expected['B'].astype(np.float64)
> > +        assert_frame_equal(result, expected)
> > +
> > +        # From a dataframe with incorrect data type for fill_value
> > +        result = df.unstack(fill_value=0.5)
> > +
> > +        rows = [[1, 3, 2, 4], [0.5, 5, 0.5, 6], [7, 0.5, 8, 0.5]]
> > +        expected = DataFrame(rows, index=list('xyz'), dtype=np.float)
> > +        expected.columns = MultiIndex.from_tuples(
> > +            [('A', 'a'), ('A', 'b'), ('B', 'a'), ('B', 'b')])
> > +        assert_frame_equal(result, expected)
> > +
> > +        # Test unstacking with date times
> > +        dv = pd.date_range('2012-01-01', periods=4).values
>
> can you add a test with timedeltas as well (similar to the datetimes test)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pydata/pandas/pull/10246/files#r50281164>.
>